impl/maven-core/src/main/java/org/apache/maven/graph/FilteredProjectDependencyGraph.java

```diff
@@ -74,12 +79,24 @@ public List<MavenProject> getSortedProjects() {
 
     @Override
     public List<MavenProject> getDownstreamProjects(MavenProject project, boolean transitive) {
-        return applyFilter(projectDependencyGraph.getDownstreamProjects(project, transitive), transitive, false);
+        Key key = new Key(project, transitive, false);
+        List<MavenProject> list = cache.get(key);
+        if (list == null) {
+            list = applyFilter(projectDependencyGraph.getDownstreamProjects(project, transitive), transitive, false);
+            cache.put(key, list);
+        }
```

Comment:
`computeIfAbsent`?

Reply:
No, that's what cause failures with #1943

Reply:
comments here why not use  `computeIfAbsenta` - can help in future